### DISCUSSION TITLE
Add document for building modmesh

### DISCUSSION THREAD

**Discussion by tychuang1211** [solvcon/modmesh#342] [General]
Adding document may help with the building process.
I'd like to start from "building modmesh on Ubuntu with [devenv](https://github.com/solvcon/devenv)".

**Comment by allen91wu** [comment]
Hi @tychuang1211 , I'm trying to build modmesh in Linux terminal. But I'm failed. Would you like to give me some direction of a solution?

**Error messages**
```
allen91wu@LAPTOP-5ULS3SR4:/mnt/c/Users/allen/Desktop/modmesh$ sudo python3 setup.py build
running build
running build_py
running build_ext
-- use PYTHON_EXECUTABLE=/usr/bin/python3
-- not use clang-tidy
ERRORflake8 not found, run pip install flake8
-- HIDE_SYMBOL: OFF
-- DEBUG_SYMBOL: ON
-- MODMESH_PROFILE: OFF
CMake Error at CMakeLists.txt:59 (find_package):
  Could not find a package configuration file provided by "pybind11" with any
  of the following names:

    pybind11Config.cmake
    pybind11-config.cmake

  Add the installation prefix of "pybind11" to CMAKE_PREFIX_PATH or set
  "pybind11_DIR" to a directory containing one of the above files.  If
  "pybind11" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
See also "/mnt/c/Users/allen/Desktop/modmesh/CMakeFiles/CMakeOutput.log".
make: *** No rule to make target '_modmesh'.  Stop.
```

**Comment by yungyuc** [comment]
Cmake cannot find pybind11.  Can you provide more information about your installation of pybind11?

**Comment by allen91wu** [comment]
After I reinstall pybind11, an error disappeared.
The following log was denoted I have successfully build the modmesh?

```
allen91wu@LAPTOP-5ULS3SR4:/mnt/c/Users/allen/Desktop/modmesh$ sudo python3 setup.py build
running build
running build_py
running build_ext
-- use PYTHON_EXECUTABLE=/usr/bin/python3
-- not use clang-tidy
-- using program '/usr/local/bin/flake8'
-- HIDE_SYMBOL: OFF
-- DEBUG_SYMBOL: ON
-- MODMESH_PROFILE: OFF
-- Found pybind11: /usr/include (found version "2.6.1" )
-- pybind11_INCLUDE_DIRS: /usr/include;/usr/include/python3.8
-- NUMPY_INCLUDE_DIR: /usr/local/lib/python3.8/dist-packages/numpy/core/include
-- CMAKE_INSTALL_INCLUDEDIR: include
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/c/Users/allen/Desktop/modmesh
make: *** No rule to make target '_modmesh'.  Stop.
```

**Comment by ianlienfa** [comment]
@yungyuc How 'bout I add the README for build process on intel Mac?

**Comment by yungyuc** [comment]
> @yungyuc How 'bout I add the README for build process on intel Mac?

Sure.  But for it to be available in `master`, it should be a comprehensive one including both Linux and Mac, the two supported platform.  It's good to start with a PR for intel mac.

**Comment by tai271828** [comment]
@ianlienfa fwiw, I have a (work-in-progress) helper script to build `modmesh` with `devenv`. The script is a bit overkilled (via launching a development environment for `SOLVCON`) but it should have demonstrated how to use `devenv` for `modmesh`. Pull requests are welcome to make the script smaller. https://gitlab.com/tai271828/toolbox/-/blob/dev/scripts/solvcon/modmesh.sh

> @yungyuc How 'bout I add the README for build process on intel Mac?

**Comment by yungyuc** [comment]
> @ianlienfa fwiw, I have a (work-in-progress) helper script to build `modmesh` with `devenv`. The script is a bit overkilled (via launching a development environment for `SOLVCON`) but it should have demonstrated how to use `devenv` for `modmesh`. Pull requests are welcome to make the script smaller. https://gitlab.com/tai271828/toolbox/-/blob/dev/scripts/solvcon/modmesh.sh
> 
> > @yungyuc How 'bout I add the README for build process on intel Mac?

It is not recommended to use `setup.py` for building:

```bash
python3 setup.py build_ext --cmake-args="-DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)" --make-args="VERBOSE=1"
```

The `setup.py` script calls `cmake` internally.  The added indirect is only to make it look like a Python extension module for future integration to a package manager for Python.

The `viewer` binary embeds a Python interpreter and will work very differently from a Python extension.

**Comment by ianlienfa** [comment]
@yungyuc @tai271828 
Some update on this thread,
I've successfully build the project on my intel Mac today, 
I think I can provide a Mac version install script like the one we see above.

**Comment by ianlienfa** [comment]
Btw, what does `devenv launch solvcon` do conceptually?
Is there some kind of remote repository maintaining the list of packages to download?
And will brew perform differently under the devenv or not?

**Comment by tai271828** [comment]
`devenv launch solvcon` is simply a quick approach for `devenv launch modmesh`, which has not been implemented. `devenv launch solvcon` tries to build a development environment by building the dependencies as much as possible. For example, if you take a closer look into `devenv launch solvcon` you will see `cmake` is compiled from source.

Will `brew` perform differently? Short (but not completely right) answer: no.
Long answer: it depends on how you manipulate `brew` and `devenv`. They `brew` and `devenv` could be set-up to help each other.

> Btw, what does `devenv launch solvcon` do conceptually? Is there some kind of remote repository maintaining the list of packages to download? And will brew perform differently under the devenv or not?

**Comment by ianlienfa** [comment]
> `devenv launch solvcon` is simply a quick approach for `devenv launch modmesh`, which has not been implemented. `devenv launch solvcon` tries to build a development environment by building the dependencies as much as possible. For example, if you take a closer look into `devenv launch solvcon` you will see `cmake` is compiled from source.
> 
> Will `brew` perform differently? Short (but not completely right) answer: no. Long answer: it depends on how you manipulate `brew` and `devenv`. They `brew` and `devenv` could be set-up to help each other.
> 
> > Btw, what does `devenv launch solvcon` do conceptually? Is there some kind of remote repository maintaining the list of packages to download? And will brew perform differently under the devenv or not?

Great, thanks for the response
I will look into the script later.

**Comment by tai271828** [comment]
Is anyone interested in [the framework of diataxis documentation style](https://diataxis.fr/), which is [officially adopted by Ubuntu](https://ubuntu.com/blog/diataxis-a-new-foundation-for-canonical-documentation)?

You may refer to [this talk of PyCon 2017](https://www.youtube.com/watch?v=qK-AnF5FJTs) or simply search on the internet with `diataxis Daniele Procida`. There are plenty of relevant (and probably more update-to-date) videos of discussion and talks.

**Comment by tai271828** [comment]
Please note the build process is in the Github Action CI. So anyone who is interested in how to build, you may refer to the CI yaml.

> `devenv launch solvcon` is simply a quick approach for `devenv launch modmesh`, which has not been implemented. `devenv launch solvcon` tries to build a development environment by building the dependencies as much as possible. For example, if you take a closer look into `devenv launch solvcon` you will see `cmake` is compiled from source.
> 
> Will `brew` perform differently? Short (but not completely right) answer: no. Long answer: it depends on how you manipulate `brew` and `devenv`. They `brew` and `devenv` could be set-up to help each other.
> 
> > Btw, what does `devenv launch solvcon` do conceptually? Is there some kind of remote repository maintaining the list of packages to download? And will brew perform differently under the devenv or not?

**Comment by yungyuc** [comment]
This was an old issue.  I converted it to conversation to give it more discussions.
